src/aliceVision/keyframe/KeyframeSelector.cpp

```diff
@@ -30,448 +32,397 @@ namespace keyframe {
-          {
-            hasKeyframe = true;
-            keyframeIndex = index;
-            minDistScore = _framesData[index].maxDistScore;
-          }
-        }
-      }
-      else
-      {
-        // use the first frame of the step
-        hasKeyframe = true;
-        keyframeIndex = frameIndex - (frameStep - 1);
-      }
-
-      // save keyframe
-      if(hasKeyframe)
-      {
-        ALICEVISION_LOG_INFO("keyframe choice : " << keyframeIndex << std::endl);
-
-        // write keyframe
-        for(std::size_t mediaIndex = 0; mediaIndex < _feeds.size(); ++mediaIndex)
-        {
-          auto& feed = *_feeds.at(mediaIndex);
+        const auto& path = mediaPaths.at(mediaIndex);
 
-          feed.goToFrame(keyframeIndex + _cameraInfos.at(mediaIndex).frameOffset);
+        // create a feed provider per mediaPaths
+        feeds.emplace_back(new dataio::FeedProvider(path));
 
-          if(_maxOutFrame == 0) // no limit of keyframes (direct evaluation)
-          {
-            feed.readImage(image, queryIntrinsics, currentImgName, hasIntrinsics);
-            writeKeyframe(image, keyframeIndex, mediaIndex);
-          }
-        }
-        _framesData[keyframeIndex].keyframe = true;
-        _keyframeIndexes.push_back(keyframeIndex);
-
-        frameIndex = keyframeIndex + _minFrameStep - 1;
-      }
-      else
-      {
-        ALICEVISION_LOG_INFO("keyframe choice : none" << std::endl);
-      }
-    }
-    ++currentFrameStep;
-  }
+        const auto& feed = *feeds.back();
 
-  if(_maxOutFrame == 0) // no limit of keyframes (evaluation and write already done)
-  {
-    return;
-  }
+        // check if feed is initialized
+        if (!feed.isInit())
+        {
+            ALICEVISION_LOG_ERROR("Cannot initialize the FeedProvider with " << path);
+            throw std::invalid_argument("Cannot while initialize the FeedProvider with " + path);
+        }
 
-  // if limited number of keyframe, select smallest sparse distance
-  {
-    std::vector< std::tuple<float, float, std::size_t> > keyframes;
+        // update minimum number of frames
+        nbFrames = std::min(nbFrames, (size_t)feed.nbFrames());
+    }
 
-    for(std::size_t i = 0; i < _framesData.size(); ++i)
+    // check if minimum number of frame is zero
+    if (nbFrames == 0)
     {
-      if(_framesData[i].keyframe)
-      {
-        keyframes.emplace_back(_framesData[i].maxDistScore, 1 / _framesData[i].avgSharpness, i);
-      }
+        ALICEVISION_LOG_ERROR("One or multiple medias can't be found or empty !");
+        throw std::invalid_argument("One or multiple medias can't be found or empty !");
     }
-    std::sort(keyframes.begin(), keyframes.end());
 
-    const std::size_t nbOutFrames = std::min(static_cast<std::size_t>(_maxOutFrame), keyframes.size());
+    int step = _minFrameStep;
+    if (_maxOutFrame > 0)
+    {
+        step = int(std::floor(double(nbFrames) / double(_maxOutFrame)));
```

Comment:
Unless `nbFrames` or `_maxOutFrame` is negative, this could be written as `step = nbFrames / _maxOutFrame` as division automatically truncates.